sn_node/src/put_validation.rs

```diff
@@ -30,48 +29,121 @@ use std::collections::{BTreeSet, HashSet};
-                    );
-                    warn!(
-                        "Record's key does not match with the value's ChunkAddress, ignoring PUT."
+                        "Record's key does not match with the value's RegisterAddress, ignoring PUT."
                     );
                     return Err(ProtocolError::RecordKeyMismatch);
                 }
+                self.validate_and_store_register(register).await
+            }
+            RecordKind::RegisterWithPayment => {
+                warn!("RegisterWith`Payment recieved but we cannot handle it yet!");
+                Ok(CmdOk::StoredSuccessfully)
+                // DO nothing yet
+            }
+        }
+    }
+
+    /// Store a prevalidated, and already paid record to the RecordStore
+    pub(crate) async fn store_prepaid_record(
+        &self,
+        record: Record,
+    ) -> Result<CmdOk, ProtocolError> {
+        let record_header = RecordHeader::from_record(&record)?;
+
+        match record_header.kind {
+            RecordKind::ChunkWithPayment | RecordKind::RegisterWithPayment => Err(
+                ProtocolError::UnexpectedRecordWithPayment(PrettyPrintRecordKey::from(record.key)),
+            ),
+            RecordKind::Chunk => {
+                let chunk = try_deserialize_record::<Chunk>(&record)?;
+
+                let record_key = record.key.clone();
+                let already_exists = self
+                    .validate_key_and_existence(&chunk.network_address(), &record_key)
+                    .await?;
+
+                if already_exists {
+                    return Ok(CmdOk::DataAlreadyPresent);
+                }
 
-                self.validate_and_store_chunk(chunk_with_payment, validate_payment)
-                    .await
+                self.store_chunk(chunk)
             }
             RecordKind::DbcSpend => {
-                let signed_spends = try_deserialize_record::<Vec<SignedSpend>>(&record)?;
+                let record_key = record.key.clone();
+                let spends = try_deserialize_record::<Vec<SignedSpend>>(&record)?;
 
-                // check if all the DbcAddresses matches with Record::key
-                if !signed_spends.iter().all(|spend| {
+                for spend in &spends {
                     let dbc_addr = DbcAddress::from_dbc_id(spend.dbc_id());
-                    record.key == NetworkAddress::from_dbc_address(dbc_addr).to_record_key()
-                }) {
-                    warn!("Record's key does not match with the value's DbcAddress, ignoring PUT.");
-                    return Err(ProtocolError::RecordKeyMismatch);
+                    let address = NetworkAddress::DbcAddress(dbc_addr);
+
+                    let already_exists = self
+                        .validate_key_and_existence(&address, &record_key)
+                        .await?;
+
+                    if already_exists {
+                        return Ok(CmdOk::DataAlreadyPresent);
+                    }
```

Comment:
Should we return early here? If a `SignedSpend` already exists locally, then we might have to make sure that the new PUt is not a double spend?
I think it is performed inside `validate_and_store_spends`.

Reply:
ahaaa, okay good spot